### ISSUE TITLE
Issue with recursing macro for recursive struct

### ISSUE THREAD

**Issue by kardeiz** [Keats/tera#202]
I'm trying to use a recursive macro on a recursive struct like so:

```jinja
{% macro formatted_label_for(subject, sep) %}
  {% if subject.parent %}
    {{ self::formatted_label_for(subject=subject.parent, sep=sep) }}
    {{sep}}
  {% endif %}
  {{subject.label}}
{% endmacro formatted_label_for %}
```

However, this crashes with:

```
thread '<unknown>' has overflowed its stack
fatal runtime error: stack overflow
```

On examination, it looks like the object itself (`subject`) is getting passed to the macro call, rather than its `parent`. I've also tried `set`ting the parent as its own variable before calling the macro, but that doesn't seem to make any difference.

Am I missing something or is this a known limitation?

**Comment by Keats** [comment]
Hmm that looks like a bug, it should be working.. Can you open a PR with a test as a reproduction?

**Comment by kardeiz** [comment]
I added a test that crashes (see PR #203 ).

Interestingly, this only happens when the macro is called inside a loop. I don't know what to make of this. Please let me know if I overlooked something. Thanks!

**Comment by Keats** [comment]
So the issue is that the code that finds the variable will see there is a forloop and will try to get the value from there rather than the macro context, resulting in pretty much an infinite loop.
That seems to point out that the current context system is a bit broken as it doesn't take into account whether a forloop is in a macro or not. For now you can rename the macro param to something other than the forloop name (`obj` currently) while I figure out something.

I'll take any idea!

**Comment by Keats** [comment]
I believe https://github.com/Keats/tera/pull/205 fixes it

**Comment by kardeiz** [comment]
Wow, that's awesome! I would have been happy with the rename workaround as a decent short-term fix. Thanks!

**Comment by Keats** [comment]
The fix is in 0.10.9
